include/boost/histogram/display.hpp

```diff
@@ -0,0 +1,212 @@
+                     const size_t& str_values_width,
+                     const size_t& external_line_shift)
+                     :
+                     str_values_{str_values}, 
+                     scale_factors_{scale_factors},
+                     lower_bounds_width_{lower_bounds_width},
+                     upper_bounds_width_{upper_bounds_width},
+                     str_values_width_{str_values_width},
+                     external_line_shift_{external_line_shift}
+                     {}
+};
+
+template <class histogram>
+extract extract_data(const histogram& h) {
+  std::string lower, upper;
+
+  auto data = indexed(h, coverage::all);
+
+  extract ex;  
+  for (auto x : data) {
+    lower = str( boost::format("%.1f") % x.bin().lower() );
+    upper = str( boost::format("%.1f") % x.bin().upper() );
+    ex.lower_bounds_.push_back(lower);
+    ex.upper_bounds_.push_back(upper);
+    ex.values_.push_back(*x);
+  }
+  return ex;    
+}
+
+std::string get_single_label(const extract& data,
+                             const unsigned int index, 
+                             const unsigned int column_width1,
+                             const unsigned int column_width2) {
+  std::string label = "";
+  char parenthesis = ' ';
+  std::string lower = data.lower_bounds_.at(index);
+  std::string upper = data.upper_bounds_.at(index);
+  if (index == data.size() - 1)
+    parenthesis = ']';
+  else
+    parenthesis = ')';
+
+  label = '[' 
+        + str(boost::format("%s") % boost::io::group(std::setw(column_width1), lower))
+        + ", "
+        + str(boost::format("%s") % boost::io::group(std::setw(column_width2), upper))
+        + parenthesis;
+  return label;
+}
+
+std::string get_single_str_value(const std::vector<std::string>& str_values,
+                                 const unsigned int index,
+                                 const unsigned int column_width) {
+  std::string str_value = str_values.at(index);
+  str_value =
+      str(boost::format("%-s") % boost::io::group(std::setw(column_width), str_value));
+  return str_value;
+}
+
+std::vector<int> calculate_scale_factors(const std::vector<int>& values) {
+  std::vector<int> scale_factors{};
+  const unsigned int longest_bin = max_bin_coefficient * histogram_width;
+
+  auto max_value = std::max_element(values.begin(), values.end());
+
+  for (const auto& x : values) {
+    int result = x * longest_bin / (*max_value);
+    scale_factors.push_back(result);
+  }
+  return scale_factors;
+}
+
+size_t get_max_width(const std::vector<std::string>& container) {
+  size_t max_length = 0;
+
+  for (const auto& line : container)
+    if (line.length() > max_length) 
+      max_length = line.length();
+  return max_length;
+}
+
+std::vector<std::string> convert_to_str_vec(const std::vector<int>& values) {
+  std::vector<std::string> string_values;
+  string_values.reserve(values.size());
+
+  std::transform(values.begin(), values.end(), std::back_inserter(string_values),
```

Comment:
Nice to see you using the STL, but again, please no allocations. You can write all this just using std::ostream.